src/engine/engine.cpp

```diff
@@ -34,24 +49,65 @@ void Engine::pop_screen() {
 
 void Engine::engine_go() {
   while (true) {
-    this->input_manager->update_input();
+    for (i32 ii = 0; ii < (int)this->updates_per_render; ii++) {
+      this->input_manager->update_input();
+
+      glClear(GL_COLOR_BUFFER_BIT);
+      this->update();
+    }
 
-    glClear(GL_COLOR_BUFFER_BIT);
-    this->update();
+    auto frame_time_start = std::chrono::high_resolution_clock::now();
 
+    this->paint();
     renderer->render();
     renderer->clear_paint_buffer();
-
     glfwSwapBuffers(this->window);
+
+    auto duration = std::chrono::duration_cast<std::chrono::microseconds>(
+        std::chrono::high_resolution_clock::now() - frame_time_start);
+
+    auto time_to_sleep_ms =
+        1000.0 / (this->FPS) - ((f32)duration.count()) / 1000.0;
+
+    if (time_to_sleep_ms > 0) {
+#ifdef __linux__
+      usleep(time_to_sleep_ms * 1000);
+#elif _WIN32
+      Sleep(time_to_sleep_ms);
+#else
+#error "Platform not supported"
+#endif
+    }
+
     if (glfwWindowShouldClose(this->window)) {
       break;
     }
   }
 }
 
 void Engine::update() {
+  InputState *input_state = this->input_manager->get_current_input_state();
+  if (input_state->slomo_down && !input_state->slomo_down_prev) {
+    this->slomo = !this->slomo;
+  }
+
+  if (this->slomo &&
+      this->updates_per_render != (f32)this->min_updates_per_render) {
+    this->updates_per_render -= 0.5;
+  }
+
+  if (!this->slomo &&
```

Comment:
Same as before - 'Now that we're working with floats & potentially non-linear slomo animations, this should be < or >, not !=, just in case we overshoot it'